### DISCUSSION TITLE
Consultas sobre SDK de Sensores de Temperatura

### DISCUSSION THREAD

**Discussion by D-M-Gonzalez** [dds-jv/foro#430] [TPA - 2024] [answered]
Buenas!

Vi una respuesta de hace 20 horas sobre este tema, pero particularmente a mi me sigue pareciendo que no esta clara la respuesta.
El SDK del sensor por lo que dice, ejecuta un método llamado onTemperatureChange cada X cantidad de tiempo. 

Por lo que entiendo, el SDK podría estar haciendo algo así ( no importa la implementación, pero para ejemplificar mejor que es lo que entiendo )

public void onTemperatureChange(Action action) {
   ...hacer cosas
   action(temperature)
}

Nosotros tenemos que implementar ese action, cosa de que onTemperatureChange la ejecute y nos pase por parámetro la nueva temperatura. Esto es, porque no tenemos manera de comunicar "para atras".

Nuestro action podría ser algo como esto:

public void executeForTemperature(double temperatura) {
   validar la temperatura nueva
}

Implementado en nuestro fachade del sensor, o algo por el estilo.
Ahora, mi duda viene, si esto que estoy pensando es lo correcto, ¿Debemos inyectar nuestro Action en el sensor verdad? ¿Cómo inyectamos nuestro action en el SDK del sensor? 
Y en caso de que no sea así, ¿Cómo hace el SDK del sensor para ejecutar nuestro action cada vez que se ejecuta onTemperatureChange? Si nosotros no podemos cambiar su implementación.

**Comment by D-M-Gonzalez** [comment]
A menos que estoy pensando, onTemperatureChange es justamente el método que inyecta la Action y queda guardado en el SDK del sensor para que se ejecute como callback cada vez que se ejecute valla a saber uno que, cada vez que cambia la temperatura.

**Comment by flbulgarelli** [comment, marked as answer]
:wave:  ¡Hola! 

El método `onTemperatureChange`....

![image](https://github.com/dds-jv/foro/assets/677436/3cbd01cf-8310-488c-850b-13060d0bb3c4)

... sirve para configurar al sensor, indicándole qué hacer **cada vez** que se cumpla la condición de cambio de temperatura. Para lograrlo, más allá de que debés asumir que ya está implementado, probablemente haga algo parecido a ésto: 

```java
public void onTemperatureChange(Action action) {
    this.accionesParaHacerAnteCambioDeTemperatura.add(action);
}
```

En otras palabras, la implementación que asumiste....

```java
public void onTemperatureChange(Action action) {
...hacer cosas
action(temperature)
}
```

... es incorrecta, porque implicaría que `onTemperaturaChange` ejecute la acción una sola vez, en ese preciso instante, y no cuando el cambio de temperatura ocurra. Tené en cuenta que la acción podría ejecutarse una vez, muchas veces o incluso nunca. Como hace el sensor para lograr que la acción se ejecute ante cada cambio de temperatura es algo que escapa al problema[^1], y que simplemente debés asumir que cuando ese cambio ocurra, tu código pueda reaccionar ante ese evento, realizando tareas y tomando decisiones propias de **tu dominio**, como por ejemplo: 

```java
class ActualizarTemperaturaDeHeladera {
      void executeForTemperature(double temperatura) { /* ..actualizar temperatura de heladera.. */ }  
}
// ...
Action unaAccion = new ActualizarTemperaturaDeHeladera(heladera)
sensor.onTemperaturaChange(unaAccion)
```

O usando la sintaxis de Java para implementar interfaces de un sólo mensaje: 

```java
sensor.onTemperaturaChange((unaTemperatura) => {
   /* ..actualizar temperatura de heladera.. */
})
```

[^1]: De todas formas, cuando estudiemos eventos, el patrón observer y la herramienta cron veremos posibles formas de implementar esta notificación automática.
